File: task2/grpc/client-java/src/test/java/com/epam/sp/PingPongClientTest.java
package com.epam.sp;

import com.epam.sp.proto.pingpong.PingPongServiceGrpc;
import com.epam.sp.proto.pingpong.Pingpong;
import com.epam.sp.server.GrpcServer;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Status;
import io.grpc.StatusRuntimeException;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

// Task 2.1.2. Create a simple Java gRPC client application that sends a request to the simple server
// (share the same proto file), gets a response, and prints it.

// Prerequisites: GrpcServer should be started
class PingPongClientTest {
    private static final GrpcServer SERVER = new GrpcServer();
    @BeforeAll
    static void setUp() throws IOException {
        SERVER.startServer();
    }

    @AfterAll
    static void cleanUp() throws InterruptedException {
        SERVER.stopServer();
    }

    @Test
    void pingPongClient_ShouldGetPongFromServer_WhenPingMessageIsSent() {
        final var input = "Ping";
        final var channel = ManagedChannelBuilder
                .forAddress("localhost", 8080)
                .usePlaintext()
                .build();

        final var stub = PingPongServiceGrpc.newBlockingStub(channel);
        final var request = Pingpong.PingRequest.newBuilder().setMessage(input).build();
        final var reply = stub.play(request);

        final var expected = "Pong";
        System.out.println("Got reply: " + reply);
        assertEquals(expected, reply.getMessage());

        channel.shutdown();
    }

    @Test
    void pingPongClient_ShouldGetStatusRuntimeException_WhenWrongMessageIsSent() {
        final var input = "Dummy";
        final var channel = ManagedChannelBuilder
                .forAddress("localhost", 8080)
                .usePlaintext()
                .build();

        final var stub = PingPongServiceGrpc.newBlockingStub(channel);
        final var request = Pingpong.PingRequest.newBuilder().setMessage(input).build();

        final var thrown = assertThrows(
                StatusRuntimeException.class,
                () -> stub.play(request),
                "Expected to get error"
        );

        assertEquals(Status.Code.INVALID_ARGUMENT, thrown.getStatus().getCode());
        assertEquals("Only 'Ping' is accepted as an input", thrown.getStatus().getDescription());
        channel.shutdown();
    }

}

File: task2/grpc/client-kotlin/src/test/kotlin/com/epam/sp/PingPongClientTest.kt
package com.epam.sp

import com.epam.sp.proto.pingpong.PingPongServiceGrpc
import com.epam.sp.proto.pingpong.Pingpong
import com.epam.sp.server.GrpcServer
import io.grpc.ManagedChannelBuilder
import io.grpc.Status
import io.grpc.StatusRuntimeException
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeAll
import kotlin.test.Test
import kotlin.test.assertFailsWith

// Task 2.1.3. Create a simple gRPC client application in any other language you like (or ask a trainer/lector for a Python example)
// that sends a request to the simple server, gets a response, and prints it. Use the same proto file.

// Prerequisites: GrpcServer should be started
class PingPongClientTest {

    @Test
    fun pingPongClient_ShouldGetPongFromServer_WhenPingMessageIsSent() {
        val input = "Ping"
        val channel = ManagedChannelBuilder
            .forAddress("localhost", 8080)
            .usePlaintext()
            .build()

        val stub = PingPongServiceGrpc.newBlockingStub(channel)
        val request = Pingpong.PingRequest.newBuilder().setMessage(input).build()
        val reply = stub.play(request)

        val expected = "Pong"
        println("Got reply: $reply")
        assertEquals(expected, reply.message)

        channel.shutdown()
    }

    @Test
    fun pingPongClient_ShouldGetGetStatusRuntimeException_WhenWrongMessageIsSent() {
        val input = "Dummy"
        val channel = ManagedChannelBuilder
            .forAddress("localhost", 8080)
            .usePlaintext()
            .build()

        val stub = PingPongServiceGrpc.newBlockingStub(channel)
        val request = Pingpong.PingRequest.newBuilder().setMessage(input).build()

        val thrown = assertFailsWith<StatusRuntimeException>(
            message = "Expected to get error",
            block = { stub.play(request) }
        )

        assertEquals(Status.Code.INVALID_ARGUMENT, thrown.status.code)
        assertEquals("Only 'Ping' is accepted as an input", thrown.status.description)
        channel.shutdown()
    }
    companion object {
        private val server = GrpcServer()
        @JvmStatic
        @BeforeAll
        fun setUp() {
            server.startServer()
        }

        @JvmStatic
        @AfterAll
        fun cleanUp() {
            server.stopServer()
        }
    }
}